types/src/v0/impls/l1.rs

```diff
@@ -64,180 +75,470 @@ impl L1BlockInfo {
+    ///
+    /// `url` must have a scheme `http` or `https`.
+    pub fn http(self, url: Url) -> L1Client {
+        L1Client::with_provider(self, Provider::new(RpcClient::Http(Http::new(url))))
+    }
+
+    /// Construct a new WebSockets client.
+    ///
+    /// `url` must have a scheme `ws` or `wss`.
+    pub async fn ws(self, url: Url) -> anyhow::Result<L1Client> {
+        Ok(L1Client::with_provider(
+            self,
+            Provider::new(RpcClient::Ws(Ws::connect(url).await?)),
+        ))
+    }
+}
+
+impl L1Client {
+    fn with_provider(opt: L1ClientOptions, mut provider: Provider<RpcClient>) -> Self {
+        let (sender, mut receiver) = async_broadcast::broadcast(opt.l1_events_channel_capacity);
+        receiver.set_await_active(false);
+        receiver.set_overflow(true);
+
+        provider.set_interval(opt.l1_polling_interval);
         Self {
-            retry_delay: Duration::from_secs(1),
-            provider,
-            events_max_block_range,
+            retry_delay: opt.l1_retry_delay,
+            provider: Arc::new(provider),
+            events_max_block_range: opt.l1_events_max_block_range,
+            state: Arc::new(Mutex::new(L1State::new(opt.l1_blocks_cache_size))),
+            sender,
+            receiver: receiver.deactivate(),
+            update_task: Default::default(),
         }
     }
+
+    /// Construct a new L1 client with the default options.
+    pub async fn new(url: Url) -> anyhow::Result<Self> {
+        L1ClientOptions::default().connect(url).await
+    }
+
+    /// Construct a new WebSockets client with the default options.
+    ///
+    /// `url` must have a scheme `ws` or `wss`.
+    pub async fn ws(url: Url) -> anyhow::Result<Self> {
+        L1ClientOptions::default().ws(url).await
+    }
+
+    /// Synchronous, infallible version of `new` for HTTP clients.
+    ///
+    /// `url` must have a scheme `http` or `https`.
+    pub fn http(url: Url) -> Self {
+        L1ClientOptions::default().http(url)
+    }
+
+    /// Start the background tasks which keep the L1 client up to date.
+    pub async fn start(&self) {
+        let mut update_task = self.update_task.0.lock().await;
+        if update_task.is_none() {
+            *update_task = Some(spawn(self.update_loop()));
+        }
+    }
+
+    /// Shut down background tasks associated with this L1 client.
+    ///
+    /// The L1 client will still be usable, but will stop updating until [`start`](Self::start) is
+    /// called again.
+    pub async fn stop(&self) {
+        if let Some(update_task) = self.update_task.0.lock().await.take() {
+            update_task.cancel().await;
+        }
+    }
+
+    fn update_loop(&self) -> impl Future<Output = ()> {
+        let rpc = self.provider.clone();
+        let retry_delay = self.retry_delay;
+        let state = self.state.clone();
+        let sender = self.sender.clone();
+
+        let span = tracing::warn_span!("L1 client update");
```

Comment:
It applies to anything but the span will only show up on WARN or higher. But now that I think about it since this span is just a tag with no extra data, it's quite cheap to log this all the time, so this could easily be `info_span!`

Reply:
er wait I had it backwards. This appears any time the log level is set to WARN or _lower_, just like a normal `tracing::warn!`. So we actually want to keep it this way, we want to see this info even if the node is running with `RUST_LOG=warn`